### PR TITLE
Speed up Patatrack CA

### PR DESCRIPTION
**Pull request by VinInn**
Release
```
           Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   14.58%  642.29ms      5000  128.46us  6.1120us  340.77us  kernel_find_ntuplets(TrackingRecHit2DSOAView const *, GPUCACell*, unsigned int const *, cms::cuda::SimpleVector<cms$
                   13.03%  573.68ms      5000  114.74us  16.704us  618.75us  gpuClustering::findClus(unsigned short const *, unsigned short const *, unsigned short const *, unsigned int const $
                    9.25%  407.24ms     10000  40.724us  2.5600us  204.64us  void kernel_BLFit<int=4>(cms::cuda::OneToManyAssoc<unsigned short, int=8, int=24576> const *, double, TrackSoAHeter$
                    8.82%  388.30ms      5000  77.659us  6.5590us  245.57us  gpuVertexFinder::vertexFinderOneKernel(ZVertexSoA*, gpuVertexFinder::WorkSpace*, int, float, float, float)
                    7.89%  347.50ms      5000  69.499us  9.1200us  316.80us  gpuPixelDoublets::getDoubletsFromHisto(GPUCACell*, unsigned int*, cms::cuda::SimpleVector<cms::cuda::VecArray<unsig$
                    5.09%  223.98ms      5000  44.795us  3.3280us  209.54us  gpuPixelDoublets::fishbone(TrackingRecHit2DSOAView const *, GPUCACell*, unsigned int const *, cms::cuda::VecArray<u$
                    5.03%  221.56ms      5000  44.311us  20.544us  174.02us  kernel_connect(cms::cuda::AtomicPairCounter*, cms::cuda::AtomicPairCounter*, TrackingRecHit2DSOAView const *, GPUCA$
                    3.99%  175.88ms      5000  35.176us  9.8230us  161.92us  gpuPixelRecHits::getHits(pixelCPEforGPU::ParamsOnGPU const *, BeamSpotPOD const *, SiPixelDigisCUDA::DeviceConstVie$
                    3.96%  174.45ms     35010  4.9820us  1.0240us  635.20us  [CUDA memcpy HtoD]
```

This PR

```
          Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   13.63%  558.73ms      5000  111.75us  15.072us  663.13us  gpuClustering::findClus(unsigned short const *, unsigned short const *, unsigned short const *, unsigned int const $
                    9.98%  409.10ms      5000  81.820us  6.0480us  193.79us  kernel_find_ntuplets(TrackingRecHit2DSOAView const *, GPUCACell*, unsigned int const *, cms::cuda::SimpleVector<cms$
                    9.86%  404.24ms     10000  40.424us  2.5600us  111.68us  void kernel_BLFit<int=4>(cms::cuda::OneToManyAssoc<unsigned short, int=8, int=24576> const *, double, TrackSoAHeter$
                    9.41%  385.77ms      5000  77.153us  6.7520us  146.88us  gpuVertexFinder::vertexFinderOneKernel(ZVertexSoA*, gpuVertexFinder::WorkSpace*, int, float, float, float)
                    8.32%  341.18ms      5000  68.235us  8.8960us  211.33us  gpuPixelDoublets::getDoubletsFromHisto(GPUCACell*, unsigned int*, cms::cuda::SimpleVector<cms::cuda::VecArray<unsig$
                    5.43%  222.45ms      5000  44.490us  3.3280us  119.14us  gpuPixelDoublets::fishbone(TrackingRecHit2DSOAView const *, GPUCACell*, unsigned int const *, cms::cuda::VecArray<u$
                    5.36%  219.69ms      5000  43.937us  20.480us  168.96us  kernel_connect(cms::cuda::AtomicPairCounter*, cms::cuda::AtomicPairCounter*, TrackingRecHit2DSOAView const *, GPUCA$
                    4.16%  170.61ms      5000  34.121us  10.367us  97.888us  gpuPixelRecHits::getHits(pixelCPEforGPU::ParamsOnGPU const *, BeamSpotPOD const *, SiPixelDigisCUDA::DeviceConstVie$
                    4.16%  170.50ms     35010  4.8700us  1.0240us  595.10us  [CUDA memcpy HtoD]
                    3.32%  135.98ms     15000  9.0650us  4.5760us  54.048us  void cms::cuda::multiBlockPrefixScan<unsigned int>(unsigned int const *, cms::cuda::multiBlockPrefixScan<unsigned i$
```

Throughput (triplets on T4)

release
  702.0 ±   1.6 ev/s
this PR
   752.4 ±   1.4 ev/s
only the first commit
  750.5 ±   1.2 ev/s

Purely technical. no regression expected. no regression observed

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
@mmusich @vmariani FYI

**Comment by mmusich** [comment]
@VinInn I am watching the all cmssw, no need to tag me explicitly.

**Comment by VinInn** [comment]
@cmsbuild , enable GPU

**Comment by slava77** [comment]
enable gpu

**Comment by slava77** [comment]
> enable gpu

@smuzaffar 
should I expect a "+1" from the bot to indicate that this hook was accepted and properly formatted, similar to what we get for the "test parameters"?

**Comment by smuzaffar** [comment]
please test
re-start with GPU

File: HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h [hunk 1]
```diff
@@ -39,6 +39,11 @@ namespace cms {
       return old;
     }
 
+    template <typename T1, typename T2>
+    T1 atomicCAS_block(T1* address, T1 compare, T2 val) {
+      return atomicCAS(address, compare, val);
```
Comment (slava77) [inline HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h:44]: for my education,
what does this and the other `_block` methods do in this PR, especially in the context of speeding things up.

**Comment by fwyzard** [comment]
They restrict the scope of the atomic operations to a single CUDA block.

For "shared" memory this is always the case.

For "global" memory this may speed up the operation, for example by not
requiring to consider the cache of other blocks

See
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions
for more details.

**Comment by slava77** [comment]
> Purely technical. no regression expected. no regression observed

there are quite a few plots in pixel tracking DQM which show small differences.
E.g. in 10824.502
wf10824 502_pv_dz

I guess either the order of entries is not exactly the same or there are some differences in results at numerical precision level.
I suppose, these are expected.
Or is it worth to investigate in more detail?

**Comment by VinInn** [comment]
GPU tracking is known NOT to be fully reproducible for a variety of reasons that have been reported earlier in various meetings. (CPU tracking would also not be reproducible if the "order" of seeds/iterations whould change (say because of high granularity threading)

**Comment by slava77** [comment]
> GPU tracking is known NOT to be fully reproducible for a variety of reasons that have been reported earlier in various meetings. (CPU tracking would also not be reproducible if the "order" of seeds/iterations whould change (say because of high granularity threading)

OK, thanks for clarifying. It was not particularly obvious that "no regression expected. no regression observed" was not a literal statement.

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34250	48c6319
- code changes look reasonable
    - some summary of the functional changes in the PR description would be useful in addition to just providing evidence that this is better as measured
- jenkins tests pass and comparisons with the baseline show no (relevant) differences in the tested CPU workflows; the GPU workfow 10824.502 shows some differences in pixel track DQM distributions, mainly at numerical precision level, attributed to some changes in the order of the produced data.

**Comment by makortel** [comment]
+heterogeneous

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h [hunk 2]
```diff
@@ -218,12 +218,13 @@ namespace gpuClustering {
               auto l = nn[k][kk];
               auto m = l + firstPixel;
               assert(m != i);
-              auto old = atomicMin(&clusterId[m], clusterId[i]);
+              auto old = atomicMin_block(&clusterId[m], clusterId[i]);
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h:221]: clusterId is in global memory, but "m" (and "i") below refers to digis on the module that is reconstructed in this block. (one module per block, one block per module)

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h [hunk 3]
```diff
@@ -350,7 +350,9 @@ __global__ void kernel_find_ntuplets(GPUCACell::Hits const *__restrict__ hhp,
     auto const &thisCell = cells[idx];
     if (thisCell.isKilled())
       continue;  // cut by earlyFishbone
-
+    // we require at least three hits...
+    if (thisCell.outerNeighbors().empty())
```
Comment (VinInn) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h:354]: simply avoids the whole lot below (in particular the function call that being recursive cannot be inlined)

File: RecoPixelVertexing/PixelVertexFinding/plugins/gpuFitVertices.h [hunk 4]
```diff
@@ -63,8 +63,8 @@ namespace gpuVertexFinder {
       assert(iv[i] >= 0);
       assert(iv[i] < int(foundClusters));
       auto w = 1.f / ezt2[i];
-      atomicAdd(&zv[iv[i]], zt[i] * w);
-      atomicAdd(&wv[iv[i]], w);
+      atomicAdd_block(&zv[iv[i]], zt[i] * w);
```
Comment (VinInn) [inline RecoPixelVertexing/PixelVertexFinding/plugins/gpuFitVertices.h:66]: vertex reconstruction is performed in a single block.

File: RecoPixelVertexing/PixelVertexFinding/plugins/gpuSortByPt2.h [hunk 5]
```diff
@@ -46,7 +46,7 @@ namespace gpuVertexFinder {
     for (auto i = threadIdx.x; i < nt; i += blockDim.x) {
       if (iv[i] > 9990)
         continue;
-      atomicAdd(&ptv2[iv[i]], ptt2[i]);
+      atomicAdd_block(&ptv2[iv[i]], ptt2[i]);
```
Comment (VinInn) [inline RecoPixelVertexing/PixelVertexFinding/plugins/gpuSortByPt2.h:49]: vertex sorting is performed in a single block

**Comment by qliphy** [comment]
+1

**Pull request merged**